sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/implementation/EventGridPublisherClientImplBuilder.java

```diff
@@ -48,22 +85,129 @@ public EventGridPublisherClientImplBuilder serializerAdapter(SerializerAdapter s
+     * @return the EventGridPublisherClientImplBuilder.
+     */
+    public EventGridPublisherClientImplBuilder httpLogOptions(HttpLogOptions httpLogOptions) {
+        this.httpLogOptions = httpLogOptions;
+        return this;
+    }
+
+    /*
+     * The retry policy that will attempt to retry failed requests, if
+     * applicable.
+     */
+    private RetryPolicy retryPolicy;
+
+    /**
+     * Sets The retry policy that will attempt to retry failed requests, if applicable.
+     *
+     * @param retryPolicy the retryPolicy value.
+     * @return the EventGridPublisherClientImplBuilder.
+     */
+    public EventGridPublisherClientImplBuilder retryPolicy(RetryPolicy retryPolicy) {
+        this.retryPolicy = retryPolicy;
+        return this;
+    }
+
+    /*
+     * The list of Http pipeline policies to add.
+     */
+    private final List<HttpPipelinePolicy> pipelinePolicies;
+
+    /**
+     * Adds a custom Http pipeline policy.
+     *
+     * @param customPolicy The custom Http pipeline policy to add.
+     * @return the EventGridPublisherClientImplBuilder.
+     */
+    public EventGridPublisherClientImplBuilder addPolicy(HttpPipelinePolicy customPolicy) {
+        pipelinePolicies.add(customPolicy);
+        return this;
+    }
+
     /**
      * Builds an instance of EventGridPublisherClientImpl with the provided parameters.
      *
      * @return an instance of EventGridPublisherClientImpl.
      */
     public EventGridPublisherClientImpl buildClient() {
+        if (apiVersion == null) {
+            this.apiVersion = "2018-01-01";
+        }
         if (pipeline == null) {
-            this.pipeline =
-                    new HttpPipelineBuilder()
-                            .policies(new UserAgentPolicy(), new RetryPolicy(), new CookiePolicy())
-                            .build();
+            this.pipeline = createHttpPipeline();
         }
         if (serializerAdapter == null) {
             this.serializerAdapter = JacksonAdapter.createDefaultSerializerAdapter();
         }
-        EventGridPublisherClientImpl client = new EventGridPublisherClientImpl(pipeline, serializerAdapter);
+        EventGridPublisherClientImpl client = new EventGridPublisherClientImpl(pipeline, serializerAdapter, apiVersion);
         return client;
     }
+
+    private HttpPipeline createHttpPipeline() {
+        Configuration buildConfiguration =
+                (configuration == null) ? Configuration.getGlobalConfiguration() : configuration;
+        if (httpLogOptions == null) {
+            httpLogOptions = new HttpLogOptions();
+        }
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+        String clientName = properties.getOrDefault(SDK_NAME, "UnknownName");
+        String clientVersion = properties.getOrDefault(SDK_VERSION, "UnknownVersion");
+        policies.add(
+                new UserAgentPolicy(httpLogOptions.getApplicationId(), clientName, clientVersion, buildConfiguration));
```

Comment:
Are we not adding `ClientOptions` in the builder  ? The applicationId in UserAgentPolicy should come from there. You can see how other SDK have done it in this issuer https://github.com/Azure/azure-sdk-for-java/issues/15107

Reply:
This is the generated builder. The convenience layer builder has the client options.